kore/src/Kore/Equation/Application.hs

```diff
@@ -168,6 +188,26 @@ attemptEquation sideCondition termLike equation =
         matchIncremental term1 term2
         & MaybeT & noteT matchError
 
+    applyAndSelectMatchResult
+        :: [MatchResult (Target variable)]
+        -> ExceptT
+            (AttemptEquationError variable)
+            simplifier
+            (Equation variable, Predicate variable)
+    applyAndSelectMatchResult [] =
+        whileApplyMatchResult
+        $ applyMatchResult
+            equationRenamed
+            (Predicate.makeFalsePredicate_, mempty)
```

Comment:
This case should be handled as if `matchIncremental` had returned `Nothing`, by throwing `MatchError`. `applyMatchResult` does not handle the `\bottom` predicate in any special way, so if unification of the arguments fails, then we would call `checkRequires` without instantiating the equation. This might account for the performance problem you found with `wasm-semantics`.

Reply:
Thanks for finding this! I hadn't thought this was wrong. I'm unable to revisit the wasm tests currently because of the various frontend issues, though.